### ISSUE TITLE
Reg. the functions AddNnet() and ScaleNnet() in nnet3/nnet-utils.h

### ISSUE THREAD

**Issue by freewym** [kaldi-asr/kaldi#970]
I have doubts regarding the function AddNnet() and ScaleNnet() in nnet3/nnet-utils.h:

For each component in an nnet, these two functions scale both the parameters in updatable components and stats in non-updatable components.  In most cases, scaling the stats in non-updatable components would not affect the results; but it would in the case where, for example, ClipGradient's self-repair is activated based on the clipping stats. 

Those clipping stats are saved in (\*to_update_in) when calling Backprop(), but (\*to_update_in) is a component in (\*NnetTrainer::delta_nnet_). In the training (\*NnetTrainer::delta_nnet_) is first added to (\*NnetTrainer::nnet_) with a scaling factor (computed from max-change and momentum) by calling AddNnet(), and then is scaled by (momentum) by calling ScaleNnet(), where the stats in non-updatable components (e.g. clipping counts) also get scaled. I don't quite understand why we need to scale the stats , as scaling the stats would lead to incorrect counting.

Does it make more sense to modify AddNnet() and ScaleNnet() so that we only scale updatable components with the given scaling factor, and scale non-updatable components with the fixed factor 1.0 (i.e., only scale parameters, not stats) ?

Yiming

**Comment by danpovey** [comment]
The reason why the stats are scaled too, is that it gives
possibly-more-sensible behavior when doing things like model-averaging.
E.g. suppose someone were to run training with --zero-component-stats=false
given to nnet3-train.  If the stats were not scaled while doing the model
averaging, then on each iteration the stats would get multiplied by the
number of jobs, and would eventually exceed the floating point range.

Also, during training it's important to scale the stats of the delta_nnet_
down (to zero, or with momentum, to scale them by a nonzero value) after
adding the stats to the neural net nnet_, because otherwise the same stats
would get added multiple times.

I know that as far as the stats for gradient-clipping are concerned, this
method may not be 100% ideal because for minibatches that had a lot of
scaling-down due to max-change the stats will get scaled down, and this
might not be what you want.  But I think it shouldn't affect anything in a
very significant way.


Dan
